acceptance tests, bugfixes & features

lukemarsden opened the pull request:
* Docker acceptance tests against real Docker.
* Fix GET bug (unable to proxy GET requests).
* Support streaming streamed responses when there are no matching post-hooks (instead of buffering).
* Stdin support, so `docker run`, `-t`, `-i`, and `-ti` all work.
* Documentation fixes and clarifications on modified semantics & limitations.

This is a clear set of improvements which is IMO worthy of a v0.0.2 release.

binocarlos commented on powerstrip/powerstrip.py:
```diff
@@ -52,7 +63,12 @@ def registerListener(self, d):
+    def dataReceived(self, data):
+        # print "DATA!", repr(data)
+        return proxy.ProxyClient.dataReceived(self, data)
+
```
perhaps we want a comment here to indicate what is happening - in my understanding we overloading this method such that data is sent directly back to the docker client?

binocarlos commented on powerstrip/powerstrip.py:
```diff
@@ -64,11 +80,16 @@ def loseWriteConnectionReason(reason):
+            def stdinHandler(data):
+                self.transport.write(data)
+            self.father.transport.protocol.dataReceived = stdinHandler
+            self.setStreamingMode(True)
```
It feels like the section above would belong nicely in a function somewhere called "handleRawStream" leaving "handleHeader" to be just a controller function

binocarlos commented on powerstrip/powerstrip.py:
```diff
@@ -170,9 +193,14 @@ def __init__(self, dockerAddr=None, dockerPort=None, dockerSocket=None,
+        elif request.requestHeaders.getRawHeaders('content-type') == ["application/tar"]:
+            # XXX We can't JSON encode binary data, so don't even try.
+            skipPreHooks = True
+            originalRequestBody = None
```
I think its reasonable to not trigger hooks for certain types of request - the change to the Readme in the PR clearly states when these exceptions are so in my view this is fair

binocarlos commented on powerstrip/powerstrip.py:
```diff
@@ -198,20 +225,28 @@ def callPreHook(result, hookURL):
                             "Body": newRequestBody,
+                            # XXX This would need a ContentType header... if we
+                            # were to support non-JSON POST bodies, like build
+                            # contexts in pre-hooks.
```
I'm not convinced we need/want to support build hooks - the ContentType header would do no harm but I don't think we should be trying to encode binary data into the packet - I think its fair to know and state the limitations - not triggering hooks for build requests is one :-)

binocarlos commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,296 @@
+        """
+        # Note that http://devnull/ should never be attempted because one
+        # should always skip a post-hook when Docker responds with a hijacked
+        # response type.
```
it shoud say here that if an attempt to connect to /devnull is made it will end with a DNS lookup error

binocarlos commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,296 @@
+              "for X in range {1..10000}; do echo \\$X; done");
+            docker wait $id >/dev/null; echo $id
+            """,
+            expectDifferentResults=True)
```
this is just so we can run a docker command not to actually compare the 2 different types - just an observation :-)

binocarlos commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,296 @@
+            self.assertNotIn("fatal", docker)
+        d.addCallback(assertions)
+        return d
+    test_run_docker_pull.skip = "only works when you are online"
```
now we have Internet we can actually run this!

lukemarsden replied:
https://github.com/ClusterHQ/powerstrip/issues/48

robhaswell replied:
Have you really deferring deleting one line to a new issue? I feel like adding half-finished tests is nasty.

lukemarsden replied:
fixed by https://github.com/ClusterHQ/powerstrip/commit/dc15847f31c6b3a099d2ea112877b55fc75f78d1 (and the newly-unskipped test passes)

binocarlos commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,296 @@
+        d.addCallback(assertions)
+        return d
+
+    def test_run_docker_pull(self):
```
also need a test for `docker pull` after an image has been downloaded which would yield a different response

lukemarsden replied:
https://github.com/ClusterHQ/powerstrip/issues/48

lukemarsden replied:
fixed (just by making the pull test clean itself up before it starts)

binocarlos commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,296 @@
+                realDockerSocket="/var/run/docker.sock")
+        self.config.read_and_parse()
+        d = CompareDockerAndPowerstrip(self,
+            "cd ../free-dockerfile; docker build .", usePTY=True)
```
the free-dockerfile "could" be in a fixtures folder

binocarlos commented on powerstrip/testtools.py:
```diff
@@ -7,6 +7,60 @@
+
+from twisted.protocols.policies import TrafficLoggingFactory
+
+DO_TRAFFIC_LOGGING = False
```
so it seems we have decided to just not do traffic logging? in which case could we not just remove that code below?

binocarlos commented on powerstrip/testtools.py:
```diff
@@ -143,3 +197,46 @@ def __init__(self, pre, post, explode, incrementBy):
+
+    def _handlePostHook(self, request, requestJson):
+        # The desired response is the entire client request
+        # payload, unmodified.
```
the response also has the response from the docker server not just the client request

binocarlos commented:
Looks good! Please address review comments and then merge...

lukemarsden commented:
Thanks! I'll let @robhaswell take a quick look at this and then merge at will.

robhaswell commented on powerstrip/powerstrip.py:
```diff
@@ -52,23 +63,37 @@ def registerListener(self, d):
-                "\r\n")
-            self._streaming = True
-            self._fireListener(Failure(NoPostHooks()))
+        # print key, "=>", value
```
Remove debugging.

lukemarsden replied:
fixed

robhaswell commented on powerstrip/powerstrip.py:
```diff
@@ -52,23 +63,37 @@ def registerListener(self, d):
+        # XXX Turns out, the build endpoint doesn't actually used chunked
+        # encoding. It just sends some JSON documents which maybe happen to
+        # line up with packet boundaries. So the following if statement is both
+        # untested and potentially never triggered in practice. :(
```
I'm gonna say the obvious things:

This should have a test case or be removed. The test case could be a demonstrator of the issue in the wild. Code which maybe works under unknown circumstances is a big big liability.

Also this knowledge should be in a docstring (.. of a test, preferably ..) instead of a code comment.

lukemarsden replied:
good shout. I've removed the code. we can always re-add it if we discover docker actually doing a transfer-encoding: chunked, which we haven't seen yet. all tests still pass.

robhaswell commented on powerstrip/powerstrip.py:
```diff
@@ -79,6 +104,7 @@ def handleResponsePart(self, buffer):
         if self._streaming:
             proxy.ProxyClient.handleResponsePart(self, buffer)
         else:
+            # print "BUFFER!", buffer
```
Remove debugging, or add a debugging mode.

lukemarsden replied:
done

robhaswell commented:
After this branch can we please start implementing acceptance tests, bugfixes and features in **separate branches**.

robhaswell commented on powerstrip/powerstrip.py:
```diff
@@ -227,9 +263,18 @@ def doneAllPrehooks(result):
+                del allRequestHeaders["transfer-encoding"]
+            # XXX Streaming the contents of the request body into memory could
+            # cause OOM issues for large build contexts POSTed through
+            # powerstrip.
```
Raise an isssue for known issue and link to the issue.

lukemarsden replied:
thanks https://github.com/ClusterHQ/powerstrip/issues/51

robhaswell commented on powerstrip/powerstrip.py:
```diff
@@ -262,6 +310,7 @@ def callPostHook(result, hookURL):
                             "Code": serverResponse["Code"],
                         },
                     }), headers={'Content-Type': ['application/json']})
+        # XXX Need to skip post-hooks for tar archives from e.g. docker export.
```
Raise an issue.

lukemarsden replied:
https://github.com/ClusterHQ/powerstrip/issues/52 - thanks!

robhaswell commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,297 @@
+        """
+        # Note that http://devnull/ should never be attempted because one
+        # should always skip a post-hook when Docker responds with a hijacked
+        # response type. If the implementation *does* attempt to connect to
```
This is unclear. What is a hijacked response type? I suspect that the docstring and method name for this test is inadequate.

As a rule, descriptions of the test should belong in the docstring.

lukemarsden replied:
84fac42763520f300e63324a8e2c0a362dadd85b

robhaswell commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,297 @@
+    def test_run_post_hook(self):
+        """
+        Test basic ``docker run`` functionality when there's a post-hook (the
+        post-hook should get skipped).
```
Why? Post-hooks should be run, surely.

robhaswell replied:
I think this applies to the following three tests.

lukemarsden replied:
Post-hooks are skipped on `attach` commands because they have streaming responses.

robhaswell replied:
So say that in the test.

lukemarsden replied:
done

robhaswell commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,297 @@
+
+
+def CompareDockerAndPowerstrip(test_case, cmd, usePTY=False,
+        expectDifferentResults=False):
```
Document `expectDifferentResults` as its behaviour is surprising. I would suggest `returnResults` is a better name.

robhaswell replied:
Also, naming scheme.

lukemarsden replied:
I think you misread - expectDifferentResults is named correctly. It always returns results.

lukemarsden replied:
I added some documentation in 8db949434e2b7b585264f5454283754d53df2e7f

robhaswell replied:
I still think this is poorly named. `expectDifferentResults`, to me, means "fail if they are the same", not "remove the fail behavior". I suggested `returnResults` because it implies that the behaviour will change significantly.

robhaswell commented on powerstrip/testtools.py:
```diff
@@ -7,6 +7,51 @@
+        self.nullPort = self.nullServer.getHost().port
+
+    def _configure(self, config_yml, dockerArgs={}, dockerOnSocket=False,
+            realDockerSocket=False, powerstripPort=0, nullAdapter=False):
```
These arguments do not match the naming convention. Raise a ticket if you don't want to fix it.

lukemarsden replied:
https://github.com/ClusterHQ/powerstrip/issues/53

robhaswell commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,297 @@
+        d = CompareDockerAndPowerstrip(self,
+            "docker run -ti ubuntu echo hello", usePTY=True)
+        def assertions((powerstrip, docker)):
+            self.assertNotIn("fatal", docker)
```
This is fragile. I suspect the successful output is much more predictable. Something like `self.assertEquals('hello', docker)`.

lukemarsden replied:
good shout! 3bda0db39e160439d6a1746d88a64bc2934980c9

robhaswell commented on powerstrip/test/test_passthru.py:
```diff
@@ -0,0 +1,297 @@
+        d = CompareDockerAndPowerstrip(self,
+            "cd ../fixtures/free-dockerfile; docker build .", usePTY=True)
+        def assertions((powerstrip, docker)):
+            self.assertNotIn("fatal", docker)
```
Could you build something small with a predictable output? E.g.

```
FROM busybox
RUN echo hello
```

Then you assert that the third (?) last line is exactly equal to `hello`.

lukemarsden replied:
This test was failing until I made fixes in the code, with respect to actually working at all with `docker build` commands. So the test does its job.

robhaswell commented on fixtures/free-dockerfile/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+RUN sleep 1
+RUN touch foo
+RUN sleep 1
+RUN touch foo
```
Is there any reason to intentionally add 5 seconds to a test run? Also this would be better built from a small image, for obvious reasons. There's an additional comment I made about this test.

lukemarsden replied:
This is useful when running `docker build` manually to observe that the output is actually streamed. This helped me fix another bug, but it's hard to test. So: https://github.com/ClusterHQ/powerstrip/issues/54

robhaswell commented on powerstrip/powerstrip.py:
```diff
@@ -227,9 +263,18 @@ def doneAllPrehooks(result):
                 rest = self.path
+            allRequestHeaders = request.getAllHeaders()
+            if ("transfer-encoding" in allRequestHeaders
+                    and allRequestHeaders["transfer-encoding"] == "chunked"):
```
`if allRequestHeaders.get("transfer-encoding") == "chunked":`

lukemarsden replied:
fixed, thanks

robhaswell commented:
I haven't reviewed this for correctness. I'm suspicious about the lack of tests for the new chunked encoding stuff. It does have a lot of mess which I would prefer not to be merged. There's also some undocumented technical debt which I've asked that you raise issues for. Please merge after addressing all my comments, thanks.

lukemarsden commented:
I've addressed all your concerns (or raised issues in one or two cases). Thanks for the thorough reviews @robhaswell @binocarlos.

```
hybrid@node001:~/powerstrip$ sudo TEST_PASSTHRU=1 trial powerstrip.test
[snip]
PASSED (skips=10, successes=52)
```

Merging, thanks!
